docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+                   &my_data,        // plaintext
+                   &my_data_size,   // plaintext_size
+                   NULL,            // additional_data
+                   NULL             // additional_data_size
+                   );
+if (result != OE_OK)
+    // Handle decryption error...
+else
+{
+    // Process my_data...
+}
+```
+
+The initial `key_info` passed to `oe_seal()` could be retrieved by a call to
+`oe_get_seal_key_by_policy()`, which takes a `seal_policy` as its only input
+parameter. In practice, an enclave may desire to encrypt different things under
+different keys. Per
+[NIST.SP800-108](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf),
+*Label* and *Context* should be among the inputs to *KDF* (*Key Derivation
+Function*). Most TEEs also support taking addtional *context* or *entropy* into
+key derivation. The existing `oe_get_seal_key_by_policy()` API however lacks
+that support and prompts for a new API to fill the gap.
+
+The new `oe_get_seal_key_info()` API is introduced with the following
+signature.
+
+```C
+oe_result_t oe_get_seal_key_info(
+    oe_seal_policy_t seal_policy,
+    uint8_t* entropy,
+    size_t entropy_size,
+    uint64_t tee_specific,
+    uint8_t** key_info,
+    size_t* key_info_size);
+```
+
+If compared to `oe_get_seal_key_by_policy()`, `oe_get_seal_key_info()` differs
+in:
+
+- The addition of `entropy`/`entropy_size` parameters - This allows *Label* or
+  *Context* (or both), or simply more entropy to be mixed into the seal key.
+- The addition of `tee_specific` parameter - This is a TEE specific parameter,
+  or `OE_SEAL_SGX` to use TEE defaults. See
+  [oe_get_seal_key_info()](#oe_get_seal_key_info()) later in this document for
+  more information.
+- The removal of `key_buffer`/`key_size` output parameters - Direct use of seal
+  keys is error prone so discouraged.
+
+The code snippet below shows how to create a `key_info` suitable for a
+provisioning enclave. It adds `PROVISION_KEY` to the attribute mask, and uses
+`PROVISION_SEAL_KEY` to exclude CPU SVN from key derivation.
+
+```C
+sgx_key_request_t *key_info;
+size_t key_info_size;
+result = oe_get_seal_key_info(OE_SEAL_POLICY_PRODUCT, NULL, 0,
+                              OE_SEAL_SGX | OE_SEALKEY_DEFAULT_FLAGSMASK |
+                              SGX_FLAGS_PROVISION_KEY,
+                              (uint8_t**)&key_info, &key_info_size);
+if (result != OE_OK)
+    // Handle errors...
+else
+    key_info->key_name = SGX_KEYSELECT_PROVISION_SEAL;
+```
+
+## Specification
+
+New files:
+
+- [include/openenclave/bits/seal.h](../../include/openenclave/bits/seal.h)
+  contains generic definitions.
+- [include/openenclave/bits/sgx/seal.h](../../include/openenclave/bits/sgx/seal.h)
+  contains SGX specific definitions.
+- [enclave/seal_gcm-aes.c](../../enclave/seal_gcm-aes.c) implements `oe_seal()`
+  and `oe_unseal()`.
+- [enclave/crypto/gcm.c](../../enclave/crypto/gcm.c) implements GCM-AES128.
```

Comment:
How is crypto-agility supported?

Reply:
Are you talking about switching to a different cipher implementation (e.g., OpenSSL vs. MbedTLS) or a different cipher algorithm (e.g., AES vs SM4)?

- The SDK picks the cipher algorithm, just like what TPM does.
- @mingweishih is working on switching between MbedTLS vs. OpenSSL. Once that's done, my thought is there'll be 2 gcm.c files, one using MbedTLS and the other OpenSSL. The user would be allowed to specify which crypto lib to link.

Reply:
+1 to Simon's response.   Key length is the main thing I'm referring to but the same argument can be made about other things.  See https://tools.ietf.org/html/rfc7696 for longer discussion.

Reply:
Only the weakest link matters in security. Given the processor's 128-bit fuse key, it couldn't add security by using AES-256 as the adversary would simply avoid the 256-bit encryption key but attack the 128-bit fuse key instead.

The idea here is to use the strongest key possible. TEEs may evolve overall time and we'd like to use stronger keys on newer platforms without concerning the users.

Moreover, unsealing can only be done on the same platform by an enclave from the same vendor. There's no reason for a single vendor to use one cipher/mode in one of its enclaves but a different cipher/mode in another. That said, the most of [RFC7696](https://tools.ietf.org/html/rfc7696) don't apply here.

Based on the 2 reasons above, I still don't believe there's a need to support m